src/batch.jl

```diff
@@ -0,0 +1,191 @@
+    end
+
+    return I, E, numevals
+end
+
+function handle_infinities(workfunc, f::BatchIntegrand, s)
+    s1, s2 = s[1], s[end]
+    if realone(s1) && realone(s2) # check for infinite or semi-infinite intervals
+        inf1, inf2 = isinf(s1), isinf(s2)
+        if inf1 || inf2
+            xtmp = f.x # buffer to store evaluation points
+            ytmp = f.y # original integrand may have different units
+            xbuf = similar(xtmp, typeof(one(eltype(f.x))))
+            ybuf = similar(ytmp, typeof(oneunit(eltype(f.y))*oneunit(s1)))
+            if inf1 && inf2 # x = t/(1-t^2) coordinate transformation
+                return workfunc(BatchIntegrand((v, t) -> begin resize!(xtmp, length(t)); resize!(ytmp, length(v));
+                                            f.f!(ytmp, xtmp .= oneunit(s1) .* t ./ (1 .- t .* t)); v .= ytmp .* (1 .+ t .* t) .* oneunit(s1) ./ (1 .- t .* t) .^ 2; end, ybuf, xbuf, f.max_batch),
+                                map(x -> isinf(x) ? (signbit(x) ? -one(x) : one(x)) : 2x / (oneunit(x)+hypot(oneunit(x),2x)), s),
+                                t -> oneunit(s1) * t / (1 - t^2))
+            end
+            let (s0,si) = inf1 ? (s2,s1) : (s1,s2) # let is needed for JuliaLang/julia#15276
+                if si < zero(si) # x = s0 - t/(1-t)
+                    return workfunc(BatchIntegrand((v, t) -> begin resize!(xtmp, length(t)); resize!(ytmp, length(v));
+                                            f.f!(ytmp, xtmp .= s0 .- oneunit(s1) .* t ./ (1 .- t)); v .= ytmp .* oneunit(s1) ./ (1 .- t) .^ 2; end, ybuf, xbuf, f.max_batch),
+                                    reverse(map(x -> 1 / (1 + oneunit(x) / (s0 - x)), s)),
+                                    t -> s0 - oneunit(s1)*t/(1-t))
+                else # x = s0 + t/(1-t)
+                    return workfunc(BatchIntegrand((v, t) -> begin resize!(xtmp, length(t)); resize!(ytmp, length(v));
+                                            f.f!(ytmp, xtmp .= s0 .+ oneunit(s1) .* t ./ (1 .- t)); v .= ytmp .* oneunit(s1) ./ (1 .- t) .^ 2; end, ybuf, xbuf, f.max_batch),
+                                    map(x -> 1 / (1 + oneunit(x) / (x - s0)), s),
+                                    t -> s0 + oneunit(s1)*t/(1-t))
+                end
+            end
+        end
+    end
+    return workfunc(f, s, identity)
+end
+
+"""
+    quadgk(f::BatchIntegrand, a,b,c...; kws...)
+
+Like [`quadgk`](@ref), but batches evaluation points for an in-place integrand to evaluate
+simultaneously. In particular, there are two differences from `quadgk`
+
+1. The function `f.f!` should be of the form `f!(y, x) = y .= f.(x)`.  That is, it writes
+   the return values of the integand `f(x)` in-place into its first argument `y`. (The
+   return value of `f!` is ignored.) See [`BatchIntegrand`](@ref) for how to define the
+   integrand.
+
+2. `f.max_batch` must be large enough to contain `4*order+2` points to evaluate two Kronrod
```

Comment:
This should be checked somewhere in `quadgk` and we should throw an `ArgumentError` if it's not the case.

Reply:
`2N*(2order+1)` is fine, as long as it's documented.